File: service/tests/faker.go
package tests

import (
	"fmt"
	"github.com/ryandem1/oar/models"
	"math/rand"
	"time"
)

// Faker is a structure that can generate randomized fake data
type Faker struct {
	seed int64
}

// newFaker will generate a new Faker object and seed the rand package.
func newFaker() *Faker {
	seed := time.Now().UnixNano()
	rand.Seed(seed)
	return &Faker{seed: seed}
}

// integer will return an integer from the rand min - max. Max is not inclusive
func (fake *Faker) integer(min int, max int) int {
	return min + rand.Intn(max-min)
}

// testSummary will return a random test summary
func (fake *Faker) testSummary() string {
	summaries := []string{
		"Ensures the /metadata endpoint is functional",
		"Checks that a valid input produces a valid output",
		"User service load test",
		"Navbar component link positive test",
		"Ensures that publishing a valid Kafka event gets consumed correctly downstream",
		"Ensures a bad input returns a correct error message",
		"Verifies that bad data does not get forwarded downstream",
		"Test user insert query is functional",
	}
	summary := summaries[fake.integer(0, len(summaries))]
	return summary
}

// testOutcome will return a random test outcome
func (fake *Faker) testOutcome() models.Outcome {
	outcomes := []models.Outcome{
		models.Passed,
		models.Failed,
	}
	outcome := outcomes[fake.integer(0, len(outcomes))]
	return outcome
}

// testAnalysis will return a random test analysis. Takes in an outcome, because valid analyses vary based on the
// outcome of the test. Pass in a nil outcome to get any random testAnalysis
func (fake *Faker) testAnalysis(outcome *models.Outcome) models.Analysis {
	var validAnalyses []models.Analysis

	if *outcome == models.Passed {
		validAnalyses = []models.Analysis{
			models.NotAnalyzed,
			models.TrueNegative,
			models.FalseNegative,
		}
	} else if *outcome == models.Failed {
		validAnalyses = []models.Analysis{
			models.NotAnalyzed,
			models.TruePositive,
			models.FalsePositive,
		}
	} else if outcome == nil {
		validAnalyses = []models.Analysis{
			models.NotAnalyzed,
			models.TruePositive,
			models.FalsePositive,
			models.TrueNegative,
			models.FalseNegative,
		}
	} else {
		panic(fmt.Errorf("error with testAnalysis parameter, must be a valid outcome or nil! Got %s", *outcome))
	}
	analysis := validAnalyses[fake.integer(0, len(validAnalyses))]
	return analysis
}

// testResolution will return a random test resolution
func (fake *Faker) testResolution() models.Resolution {
	analyses := []models.Resolution{
		models.Unresolved,
		models.NotNeeded,
		models.QuickFix,
		models.TicketCreated,
		models.TestFixed,
		models.TestDisabled,
		models.KnownIssue,
	}
	resolution := analyses[fake.integer(0, len(analyses))]
	return resolution
}

// testDoc will generate a random testDoc. For simplicity, these docs are finite and hard-coded.
func (fake *Faker) testDoc() map[string]any {
	docs := []map[string]any{
		{
			"app":   "user-service",
			"type":  "integration",
			"owner": "Patrick Star",
			"testPayload": map[string]any{
				"id":            1,
				"username":      "someUser48",
				"accountStatus": "lock",
			},
			"testResponse": map[string]any{
				"responseCode": 200,
				"responseBody": nil,
			},
		},
		{
			"owner":         "Sandy Cheeks",
			"type":          "UI",
			"browsers":      []string{"chrome", "firefox", "edge"},
			"screenshotURL": "https://some-s3-bucket-that-doesnt-exist.com/714029473432412",
		},
		{
			"owner":   "Squidward Tentacles",
			"type":    "load",
			"maxRPS":  300,
			"service": "application-service",
			"samplePayloads": []map[string]any{
				{
					"app_id": "47324033",
					"status": "APPROVED",
				},
				{
					"app_id": "9948302",
					"status": "REJECTED",
				},
			},
			"runtime": "10m",
			"latency (ms)": map[string]float32{
				"p50": 254.33,
				"p75": 332.45,
				"p95": 501.99,
				"p99": 676.51,
			},
		},
	}
	doc := docs[fake.integer(0, len(docs))]
	return doc
}

// test will generator a random, valid models.Test object and return a pointer to it.
func (fake *Faker) test() *models.Test {
	outcome := fake.testOutcome()
	analysis := fake.testAnalysis(&outcome)
	resolution := fake.testResolution()

	test := &models.Test{
		ID:         int64(fake.integer(1, 1000000)),
		Summary:    fake.testSummary(),
		Outcome:    outcome,
		Analysis:   analysis,
		Resolution: resolution,
		Created:    time.Now(),
		Modified:   time.Now(),
		Doc:        fake.testDoc(),
	}
	return test
}

// multiple will call a specific fakeMethod function n times and return the results as a slice
func multiple[T any](n int, fakerMethod func() T) []T {
	sl := make([]T, n)
	for i := 0; i < n; i++ {
		sl[i] = fakerMethod()
	}
	return sl
}

File: enrich-ui/src/lib/faker.ts
export const fakeTests = [
	{
		id: 20,
		summary: 'Ensures the /metadata endpoint is functional',
		outcome: 'Failed',
		analysis: 'TruePositive',
		resolution: 'NotNeeded',
		created: '2023-05-05T04:30:03.531334Z',
		modified: '2023-05-05T04:30:03.541662Z',
		doc: {
			browsers: ['chrome', 'firefox', 'edge'],
			owner: 'Sandy Cheeks',
			screenshotURL: 'https://some-s3-bucket-that-doesnt-exist.com/714029473432412',
			'test left merge field': 'different value, different type',
			type: 'UI'
		}
	},
	{
		id: 14,
		summary: 'User service load test',
		outcome: 'Passed',
		analysis: 'TrueNegative',
		resolution: 'TicketCreated',
		created: '2023-05-05T04:30:03.458693Z',
		modified: '2023-05-05T04:30:03.458693Z',
		doc: {
			browsers: ['chrome', 'firefox', 'edge'],
			owner: 'Sandy Cheeks',
			screenshotURL: 'https://some-s3-bucket-that-doesnt-exist.com/714029473432412',
			type: 'UI'
		}
	},
	{
		id: 13,
		summary: 'Ensures a bad input returns a correct error message',
		outcome: 'Failed',
		analysis: 'TruePositive',
		resolution: 'KnownIssue',
		created: '2023-05-05T04:30:03.457544Z',
		modified: '2023-05-05T04:30:03.457544Z',
		doc: {
			'latency (ms)': {
				p50: 254.33,
				p75: 332.45,
				p95: 501.99,
				p99: 676.51
			},
			maxRPS: 300,
			owner: 'Squidward Tentacles',
			runtime: '10m',
			samplePayloads: [
				{
					app_id: '47324033',
					status: 'APPROVED'
				},
				{
					app_id: '9948302',
					status: 'REJECTED'
				}
			],
			service: 'application-service',
			type: 'load'
		}
	},
	{
		id: 12,
		summary: 'Ensures that publishing a valid Kafka event gets consumed correctly downstream',
		outcome: 'Passed',
		analysis: 'FalseNegative',
		resolution: 'TestFixed',
		created: '2023-05-05T04:30:03.456243Z',
		modified: '2023-05-05T04:30:03.456243Z',
		doc: {
			'latency (ms)': {
				p50: 254.33,
				p75: 332.45,
				p95: 501.99,
				p99: 676.51
			},
			maxRPS: 300,
			owner: 'Squidward Tentacles',
			runtime: '10m',
			samplePayloads: [
				{
					app_id: '47324033',
					status: 'APPROVED'
				},
				{
					app_id: '9948302',
					status: 'REJECTED'
				}
			],
			service: 'application-service',
			type: 'load'
		}
	},
	{
		id: 11,
		summary: 'Ensures that publishing a valid Kafka event gets consumed correctly downstream',
		outcome: 'Failed',
		analysis: 'TruePositive',
		resolution: 'TestDisabled',
		created: '2023-05-05T04:30:03.454872Z',
		modified: '2023-05-05T04:30:03.454872Z',
		doc: {
			app: 'user-service',
			owner: 'Patrick Star',
			testPayload: {
				accountStatus: 'lock',
				id: 1,
				username: 'someUser48'
			},
			testResponse: {
				responseBody: null,
				responseCode: 200
			},
			type: 'integration'
		}
	},
	{
		id: 10,
		summary: 'Ensures that publishing a valid Kafka event gets consumed correctly downstream',
		outcome: 'Failed',
		analysis: 'FalsePositive',
		resolution: 'TicketCreated',
		created: '2023-05-05T04:30:03.451906Z',
		modified: '2023-05-05T04:30:03.451906Z',
		doc: {
			app: 'user-service',
			owner: 'Patrick Star',
			testPayload: {
				accountStatus: 'lock',
				id: 1,
				username: 'someUser48'
			},
			testResponse: {
				responseBody: null,
				responseCode: 200
			},
			type: 'integration'
		}
	},
	{
		id: 9,
		summary: 'Test user insert query is functional',
		outcome: 'Failed',
		analysis: 'FalsePositive',
		resolution: 'TestDisabled',
		created: '2023-05-05T04:30:03.365358Z',
		modified: '2023-05-05T04:30:03.365358Z',
		doc: {
			'latency (ms)': {
				p50: 254.33,
				p75: 332.45,
				p95: 501.99,
				p99: 676.51
			},
			maxRPS: 300,
			owner: 'Squidward Tentacles',
			runtime: '10m',
			samplePayloads: [
				{
					app_id: '47324033',
					status: 'APPROVED'
				},
				{
					app_id: '9948302',
					status: 'REJECTED'
				}
			],
			service: 'application-service',
			type: 'load'
		}
	},
	{
		id: 8,
		summary: 'Verifies that bad data does not get forwarded downstream',
		outcome: 'Passed',
		analysis: 'NotAnalyzed',
		resolution: 'TestFixed',
		created: '2023-05-05T04:30:03.338032Z',
		modified: '2023-05-05T04:30:03.338032Z',
		doc: {
			'latency (ms)': {
				p50: 254.33,
				p75: 332.45,
				p95: 501.99,
				p99: 676.51
			},
			maxRPS: 300,
			owner: 'Squidward Tentacles',
			runtime: '10m',
			samplePayloads: [
				{
					app_id: '47324033',
					status: 'APPROVED'
				},
				{
					app_id: '9948302',
					status: 'REJECTED'
				}
			],
			service: 'application-service',
			type: 'load'
		}
	},
	{
		id: 7,
		summary: 'Ensures the /metadata endpoint is functional',
		outcome: 'Failed',
		analysis: 'TruePositive',
		resolution: 'NotNeeded',
		created: '2023-05-05T04:30:03.312466Z',
		modified: '2023-05-05T04:30:03.312466Z',
		doc: {
			'latency (ms)': {
				p50: 254.33,
				p75: 332.45,
				p95: 501.99,
				p99: 676.51
			},
			maxRPS: 300,
			owner: 'Squidward Tentacles',
			runtime: '10m',
			samplePayloads: [
				{
					app_id: '47324033',
					status: 'APPROVED'
				},
				{
					app_id: '9948302',
					status: 'REJECTED'
				}
			],
			service: 'application-service',
			type: 'load'
		}
	},
	{
		id: 6,
		summary: 'Ensures a bad input returns a correct error message',
		outcome: 'Passed',
		analysis: 'NotAnalyzed',
		resolution: 'QuickFix',
		created: '2023-05-05T04:30:03.286965Z',
		modified: '2023-05-05T04:30:03.286965Z',
		doc: {
			'latency (ms)': {
				p50: 254.33,
				p75: 332.45,
				p95: 501.99,
				p99: 676.51
			},
			maxRPS: 300,
			owner: 'Squidward Tentacles',
			runtime: '10m',
			samplePayloads: [
				{
					app_id: '47324033',
					status: 'APPROVED'
				},
				{
					app_id: '9948302',
					status: 'REJECTED'
				}
			],
			service: 'application-service',
			type: 'load'
		}
	},
	{
		id: 5,
		summary: 'Ensures the /metadata endpoint is functional',
		outcome: 'Passed',
		analysis: 'TrueNegative',
		resolution: 'TestDisabled',
		created: '2023-05-05T04:30:03.263148Z',
		modified: '2023-05-05T04:30:03.263148Z',
		doc: {
			'latency (ms)': {
				p50: 254.33,
				p75: 332.45,
				p95: 501.99,
				p99: 676.51
			},
			maxRPS: 300,
			owner: 'Squidward Tentacles',
			runtime: '10m',
			samplePayloads: [
				{
					app_id: '47324033',
					status: 'APPROVED'
				},
				{
					app_id: '9948302',
					status: 'REJECTED'
				}
			],
			service: 'application-service',
			type: 'load'
		}
	},
	{
		id: 4,
		summary: 'Ensures the /metadata endpoint is functional',
		outcome: 'Passed',
		analysis: 'NotAnalyzed',
		resolution: 'TicketCreated',
		created: '2023-05-05T04:30:03.137204Z',
		modified: '2023-05-05T04:30:03.166584Z',
		doc: {
			app: 'user-service',
			owner: 'Patrick Star',
			testPayload: {
				accountStatus: 'lock',
				id: 1,
				username: 'someUser48'
			},
			testResponse: {
				responseBody: null,
				responseCode: 200
			},
			type: 'integration'
		}
	},
	{
		id: 1,
		summary: 'Ensures that publishing a valid Kafka event gets consumed correctly downstream',
		outcome: 'Passed',
		analysis: 'NotAnalyzed',
		resolution: 'NotNeeded',
		created: '2023-05-05T04:30:03.005655Z',
		modified: '2023-05-05T04:30:03.005655Z',
		doc: {
			browsers: ['chrome', 'firefox', 'edge'],
			owner: 'Sandy Cheeks',
			screenshotURL: 'https://some-s3-bucket-that-doesnt-exist.com/714029473432412',
			type: 'UI'
		}
	}
];

export function selectRandomItem<T>(items: T[]): T {
	return items[Math.floor(Math.random() * items.length)];
}
